### ISSUE TITLE
Grails 3.3.8: lazy loading with non-default datasource not working

### ISSUE THREAD

**Issue by romaicus** [grails/grails-data-mapping#1198]
### Task List

- [YES ] Steps to reproduce provided
- [ YES] Stacktrace (if present) provided
- [ YES] Example that reproduces the problem uploaded to Github
- [ YES] Full description of the issue provided (see below)

## Description
we have two domain entities "Author" and "Book", they have bi-directional One-to-Many relationship.

we have two data sources defined in "Configuration/application.yml" that are "dataSource" and "dataSource2".

when "Author" and "Book" domain entities are mapped to "dataSource2" we get lazy loading error,

but when same entities are mapped to default "dataSource" we do not get lazy loading error - also for that we do no need to have "Initialization/myapp/PersistenceConfig" configuration file that sets-up "GrailsOpenSessionInViewInterceptor" 

it seems like lazy loading does not work with non-default data sources.

## Current Setup

1. we are using H2 in-memory database
2. there are two data sources defined, one is default 'dataSource' and another one is "dataSource2"
3. we have defined "Initialization/myapp/PersistenceConfig" configuration file that creates "GrailsOpenSessionInViewInterceptor" for non default data source "dataSource2"
4. Login using username: admin and password: admin
4. Currently "Author" and "Book" both entities are mapped to default data source (because no datasource is defined on the entities)
5. TestController loads "Author" entity using "createCriteria"."list(){}" and associated view "index.gsp" iterates over the list of "Author" and associated "Book"
6. This current setup works as per expectations using Lazy loading

## Steps to Reproduce
1. To reproduce simple uncomment "static mapping" in "Author" and "Book" entities
2. this will map both the entities to non-default data source that is "dataSource2"
3. Run application using grails run-app or ./grailsw run-app (grant execute permission to grailsw), use admin/admin as loing credentials and execute "TestController" where list of "Author" entity is queried and passed to the corresponding gsp that "Views/test/index.gsp"
3. when index.gsp iterates over the list of "Author" and associated list of "Book" it throws "PooledConnection has already been closed"

Note: 
1. If Configuration file "Initialization/PersistenceConfig" that sets-up "GrailsOpenSessionInViewInterceptor" for data source "dataSource2" is not defined then we get
"failed to lazily initialize a collection of role: myapp.Author.books, could not initialize proxy - no Session"

2. if we iterate over the Author's list of books in the controller which is marked "@@Transactional" this populated the Author's books but it is not accepted
because we want this to work in gsp to lazily load associated entities using OSIV.

	
### Expected Behavior
1. it should not matter what data source is defined for the entities whether default or non-default, lazy loading should work.

Tell us what should happen

#### it seems like lazy loading does not work with non-default data sources.
we have two domain entities "Author" and "Book", they have bi-directional One-to-Many relationship.

we have two data sources defined in "Configuration/application.yml" that are "dataSource" and "dataSource2".

when "Author" and "Book" domain entities are mapped to "dataSource2" we get lazy loading error,

but when same entities are mapped to default "dataSource" we do not get lazy loading error - also for that we do no need to have "Initialization/myapp/PersistenceConfig" configuration file that sets-up "GrailsOpenSessionInViewInterceptor"

Tell us what happens instead
1. it should not matter what data source is defined for the entities whether default or non-default, lazy loading should work.

### Environment Information

- **Operating System**: 
ubuntu 18.04

- **Grails Version:** 
3.3.8

- **JDK Version:** 
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-1ubuntu0.18.04.1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)

- **Container Version (If Applicable):**
N/A

### Example Application
https://github.com/romaicus/grails-nondefault-datasource-lazy-loading-fail.git

**Comment by graemerocher** [comment]
This is the expected behaviour. TransactionManager chaining is disabled by default as it adds unwanted overhead for the case where you don't require all data sources. See the upgrade notes http://docs.grails.org/latest/guide/upgrading.html

Section `TransactionManager Chaining for Multiple Data Sources Disabled by Default`

**Comment by romaicus** [comment]
Thanks Graeme, I've just noticed that the properties I had set to enable the chainedTransactionManager were different than in the documentation, but correcting them does not make any difference. 

The following is an extract from the updated project's application.yml.

`    
grails:
    ...
    transaction:
        chainedTransactionManager:
            enabled: true
            blacklistPattern: '.*'
`
We still get connection closed exceptions in the view. 

i have committed updated application.yml
[](https://github.com/romaicus/grails-nondefault-datasource-lazy-loading-fail/commit/4b196315ba462cebc9f875e2379595c517a0ca28)

Is there something else we need to configure?
